File: scripts/rakuten-daily-extract.ts
/**
 * Rakuten RMS daily sales & access data extraction script.
 *
 * Required environment variables:
 *   RAKUTEN_SERVICE_SECRET  - RMS service secret
 *   RAKUTEN_LICENSE_KEY     - RMS license key
 *
 * Optional:
 *   RAKUTEN_OUTPUT_DIR      - directory to write CSVs (default: ./data/rakuten)
 *   RAKUTEN_TARGET_DATE     - YYYY-MM-DD to extract (default: yesterday JST)
 */

import * as fs from 'fs';
import * as path from 'path';

// ── Configuration ─────────────────────────────────────────────────────────────

const SERVICE_SECRET = process.env.RAKUTEN_SERVICE_SECRET ?? '';
const LICENSE_KEY    = process.env.RAKUTEN_LICENSE_KEY    ?? '';
const OUTPUT_DIR     = process.env.RAKUTEN_OUTPUT_DIR     ?? path.join(process.cwd(), 'data', 'rakuten');

const RMS_ORDER_SEARCH  = 'https://api.rms.rakuten.co.jp/es/2.0/order/searchOrder';
const RMS_ORDER_DETAIL  = 'https://api.rms.rakuten.co.jp/es/2.0/order/getOrder';
// Item access statistics are available via the RMS Statistics download API.
// Direct per-item access API is not part of the standard RMS REST API; the
// closest is the CSV export from RMS > アクセス・売上分析 which must be
// downloaded manually or via the Statistics CSV endpoint below.
const RMS_STATS_BASE    = 'https://api.rms.rakuten.co.jp/es/1.0/report/order';

// ── Helpers ───────────────────────────────────────────────────────────────────

function authHeader(): string {
  return `ESA ${Buffer.from(`${SERVICE_SECRET}:${LICENSE_KEY}`).toString('base64')}`;
}

/** Returns yesterday's date range as JST ISO-8601 strings. */
function yesterdayJST(overrideDate?: string): { date: string; from: string; to: string } {
  let base: Date;
  if (overrideDate) {
    base = new Date(`${overrideDate}T00:00:00+09:00`);
  } else {
    const now = new Date();
    // Shift to JST (UTC+9) and subtract one day
    const jstNow = new Date(now.getTime() + 9 * 60 * 60 * 1000);
    jstNow.setUTCDate(jstNow.getUTCDate() - 1);
    base = new Date(Date.UTC(jstNow.getUTCFullYear(), jstNow.getUTCMonth(), jstNow.getUTCDate()));
  }
  const y = base.getUTCFullYear();
  const m = String(base.getUTCMonth() + 1).padStart(2, '0');
  const d = String(base.getUTCDate()).padStart(2, '0');
  return {
    date: `${y}-${m}-${d}`,
    from: `${y}-${m}-${d}T00:00:00+0900`,
    to:   `${y}-${m}-${d}T23:59:59+0900`,
  };
}

function csv(rows: string[][]): string {
  return rows.map(r => r.map(v => `"${String(v ?? '').replace(/"/g, '""')}"`).join(',')).join('\n') + '\n';
}

function ensureDir(dir: string) {
  fs.mkdirSync(dir, { recursive: true });
}

// ── API calls ─────────────────────────────────────────────────────────────────

async function searchOrderIds(from: string, to: string): Promise<string[]> {
  const res = await fetch(RMS_ORDER_SEARCH, {
    method: 'POST',
    headers: {
      Authorization: authHeader(),
      'Content-Type': 'application/json; charset=utf-8',
    },
    body: JSON.stringify({
      dateType: 2,        // 2 = 注文日
      startDatetime: from,
      endDatetime: to,
    }),
  });

  if (!res.ok) {
    const body = await res.text().catch(() => '');
    throw new Error(`searchOrder HTTP ${res.status}: ${body}`);
  }

  const json: any = await res.json();
  if (json.MessageModelList?.some((m: any) => m.messageType === 'ERROR')) {
    const msgs = json.MessageModelList.map((m: any) => m.message).join('; ');
    throw new Error(`searchOrder API error: ${msgs}`);
  }

  return json.orderNumberList ?? [];
}

async function fetchOrderDetails(orderNumbers: string[]): Promise<any[]> {
  if (orderNumbers.length === 0) return [];

  // RMS API accepts up to 100 orders per request
  const chunks: string[][] = [];
  for (let i = 0; i < orderNumbers.length; i += 100) {
    chunks.push(orderNumbers.slice(i, i + 100));
  }

  const results: any[] = [];
  for (const chunk of chunks) {
    const res = await fetch(RMS_ORDER_DETAIL, {
      method: 'POST',
      headers: {
        Authorization: authHeader(),
        'Content-Type': 'application/json; charset=utf-8',
      },
      body: JSON.stringify({ orderNumberList: chunk }),
    });

    if (!res.ok) {
      const body = await res.text().catch(() => '');
      throw new Error(`getOrder HTTP ${res.status}: ${body}`);
    }

    const json: any = await res.json();
    results.push(...(json.orderModelList ?? []));
  }

  return results;
}

// ── Aggregation ───────────────────────────────────────────────────────────────

interface ProductSalesSummary {
  itemId: string;
  itemName: string;
  manageNumber: string;
  unitCount: number;
  totalAmount: number;
  orderCount: number;
}

function aggregateSalesByProduct(orders: any[]): ProductSalesSummary[] {
  const map = new Map<string, ProductSalesSummary>();

  for (const order of orders) {
    // Skip cancelled orders
    if (order.orderProgress === 900) continue;

    for (const pkg of order.PackageModelList ?? []) {
      for (const item of pkg.ItemModelList ?? []) {
        const key = item.itemId ?? item.manageNumber ?? item.itemName ?? 'unknown';
        const existing = map.get(key);
        if (existing) {
          existing.unitCount   += item.units ?? 0;
          existing.totalAmount += (item.price ?? 0) * (item.units ?? 0);
          existing.orderCount  += 1;
        } else {
          map.set(key, {
            itemId:        item.itemId        ?? '',
            itemName:      item.itemName      ?? '',
            manageNumber:  item.manageNumber  ?? '',
            unitCount:     item.units         ?? 0,
            totalAmount:   (item.price ?? 0) * (item.units ?? 0),
            orderCount:    1,
          });
        }
      }
    }
  }

  return Array.from(map.values()).sort((a, b) => b.totalAmount - a.totalAmount);
}

// ── Main ──────────────────────────────────────────────────────────────────────

type ExtractionResult =
  | { ok: true;  date: string; salesRows: string[][]; note: string }
  | { ok: false; date: string; error: string };

async function run(): Promise<ExtractionResult> {
  const { date, from, to } = yesterdayJST(process.env.RAKUTEN_TARGET_DATE);
  console.log(`\n=== 楽天RMS 日次データ抽出 (${date}) ===\n`);

  // ── Credential check ─────────────────────────────────────────────────────
  if (!SERVICE_SECRET || !LICENSE_KEY) {
    const msg =
      'Rakuten RMS credentials not found. ' +
      'Set RAKUTEN_SERVICE_SECRET and RAKUTEN_LICENSE_KEY environment variables.';
    console.error(`[ERROR] ${msg}`);
    return { ok: false, date, error: msg };
  }

  try {
    // ── 1. Sales by product ────────────────────────────────────────────────
    console.log('1. 注文検索中...');
    const orderIds = await searchOrderIds(from, to);
    console.log(`   → ${orderIds.length} 件の注文を取得`);

    let salesRows: string[][] = [];
    let accessNote = '';

    if (orderIds.length === 0) {
      salesRows = [['商品ID', '商品名', '管理番号', '受注件数', '販売数量', '売上金額（円）']];
      console.log('   注文なし');
    } else {
      console.log('2. 注文詳細取得中...');
      const orders = await fetchOrderDetails(orderIds);
      console.log(`   → ${orders.length} 件の注文詳細を取得`);

      const summary = aggregateSalesByProduct(orders);

      salesRows = [
        ['商品ID', '商品名', '管理番号', '受注件数', '販売数量', '売上金額（円）'],
        ...summary.map(s => [
          s.itemId,
          s.itemName,
          s.manageNumber,
          String(s.orderCount),
          String(s.unitCount),
          String(s.totalAmount),
        ]),
      ];
      console.log(`   → ${summary.length} 商品に集計`);
    }

    // ── 2. Access count by product ─────────────────────────────────────────
    // The per-product access count is NOT available via the standard RMS Order
    // REST API. It is provided as a downloadable CSV report in RMS under:
    //   店舗設定 > アクセス・売上分析 > 商品別アクセス数
    // or via the RMS Statistics CSV download endpoint (separate contract).
    // A placeholder row is included to surface this gap.
    accessNote =
      '商品別アクセス数はRMS標準REST APIでは取得不可。' +
      'RMS管理画面「アクセス・売上分析」からCSVダウンロードが必要。';
    console.log(`\n[INFO] アクセス数: ${accessNote}`);

    return { ok: true, date, salesRows, note: accessNote };
  } catch (err: any) {
    const msg = err?.message ?? String(err);
    console.error(`[ERROR] ${msg}`);
    return { ok: false, date, error: msg };
  }
}

// ── Write output ──────────────────────────────────────────────────────────────

(async () => {
  const result = await run();

  ensureDir(OUTPUT_DIR);
  const ts = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);

  if (!result.ok) {
    // Write a status file so the routine leaves a paper trail
    const statusPath = path.join(OUTPUT_DIR, `extraction-status-${result.date}.txt`);
    const statusText =
      `楽天RMS 日次データ抽出\n` +
      `対象日: ${result.date}\n` +
      `実行日時: ${new Date().toISOString()}\n` +
      `ステータス: 失敗\n` +
      `理由: ${result.error}\n` +
      `\n` +
      `必要な環境変数:\n` +
      `  RAKUTEN_SERVICE_SECRET  - RMSサービスシークレット\n` +
      `  RAKUTEN_LICENSE_KEY     - RMSライセンスキー\n` +
      `\n` +
      `取得を設定するには:\n` +
      `  1. 楽天RMS > システム・WEB API設定 でAPIキーを取得\n` +
      `  2. 上記の環境変数を .env に設定\n` +
      `  3. npx tsx scripts/rakuten-daily-extract.ts を再実行\n`;

    fs.writeFileSync(statusPath, statusText, 'utf-8');
    console.log(`\nステータスファイル: ${statusPath}`);
    process.exit(1);
  }

  // Sales CSV
  const salesPath = path.join(OUTPUT_DIR, `sales-by-product-${result.date}.csv`);
  fs.writeFileSync(salesPath, csv(result.salesRows), 'utf-8');
  console.log(`\n売上CSV: ${salesPath}`);

  // Access count placeholder CSV
  const accessRows = [
    ['商品ID', '商品名', 'アクセス数', '備考'],
    ['', '', '', result.note],
  ];
  const accessPath = path.join(OUTPUT_DIR, `access-by-product-${result.date}.csv`);
  fs.writeFileSync(accessPath, csv(accessRows), 'utf-8');
  console.log(`アクセスCSV: ${accessPath}`);

  // Combined summary TXT
  const summaryPath = path.join(OUTPUT_DIR, `summary-${result.date}.txt`);
  const salesCount  = result.salesRows.length - 1; // minus header
  const totalRevenue = result.salesRows.slice(1)
    .reduce((sum, r) => sum + (parseInt(r[5], 10) || 0), 0);

  const summaryText =
    `楽天RMS 日次データ サマリー\n` +
    `${'='.repeat(40)}\n` +
    `対象日    : ${result.date}\n` +
    `実行日時  : ${new Date().toISOString()}\n` +
    `${'─'.repeat(40)}\n` +
    `[売上]\n` +
    `  商品種別数 : ${salesCount} 商品\n` +
    `  合計売上   : ¥${totalRevenue.toLocaleString()}\n` +
    `\n` +
    `[アクセス数]\n` +
    `  ${result.note}\n` +
    `${'─'.repeat(40)}\n` +
    `出力ファイル:\n` +
    `  売上CSV  : ${salesPath}\n` +
    `  アクセスCSV: ${accessPath}\n`;

  fs.writeFileSync(summaryPath, summaryText, 'utf-8');
  console.log(`サマリー  : ${summaryPath}`);
  console.log('\n抽出完了');
})();

File: scripts/rakuten_daily_extract.py
#!/usr/bin/env python3
"""
Rakuten RMS daily sales and traffic data extractor.

Reads credentials from environment variables:
  RAKUTEN_SERVICE_SECRET  - RMS service secret
  RAKUTEN_LICENSE_KEY     - RMS license key

Outputs two CSV files per run:
  data/rakuten_sales_YYYYMMDD.csv   - sales by product (from Order API v2)
  data/rakuten_traffic_YYYYMMDD.csv - access counts by product (from Navigate API)
"""

import os
import sys
import json
import base64
import datetime
import csv
import urllib.request
import urllib.error
import urllib.parse
from pathlib import Path

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

BASE_URL = "https://api.rms.rakuten.co.jp/es"
DATA_DIR = Path(__file__).parent.parent / "data"

# Target date: yesterday (JST)
JST = datetime.timezone(datetime.timedelta(hours=9))
TODAY_JST = datetime.datetime.now(JST).date()
TARGET_DATE = TODAY_JST - datetime.timedelta(days=1)
DATE_STR = TARGET_DATE.strftime("%Y%m%d")
DATE_LABEL = TARGET_DATE.strftime("%Y-%m-%d")

# Order statuses to include (all non-cancelled/non-returned orders)
ORDER_STATUS_LIST = [300, 400, 500, 600, 700, 800, 900]


# ---------------------------------------------------------------------------
# Auth helpers
# ---------------------------------------------------------------------------

def build_auth_header(service_secret: str, license_key: str) -> str:
    token = base64.b64encode(f"{service_secret}:{license_key}".encode()).decode()
    return f"ESA {token}"


def get_credentials() -> tuple[str, str]:
    secret = os.environ.get("RAKUTEN_SERVICE_SECRET", "")
    key = os.environ.get("RAKUTEN_LICENSE_KEY", "")
    return secret, key


# ---------------------------------------------------------------------------
# HTTP helpers
# ---------------------------------------------------------------------------

def api_post(path: str, payload: dict, auth: str) -> dict:
    url = BASE_URL + path
    data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        headers={
            "Authorization": auth,
            "Content-Type": "application/json; charset=utf-8",
        },
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=30) as resp:
        return json.loads(resp.read().decode("utf-8"))


def api_get(path: str, params: dict, auth: str) -> dict:
    url = BASE_URL + path
    if params:
        url += "?" + urllib.parse.urlencode(params)
    req = urllib.request.Request(
        url,
        headers={"Authorization": auth},
        method="GET",
    )
    with urllib.request.urlopen(req, timeout=30) as resp:
        return json.loads(resp.read().decode("utf-8"))


# ---------------------------------------------------------------------------
# Sales extraction (Order API v2)
# ---------------------------------------------------------------------------

def fetch_orders(auth: str) -> list[dict]:
    """Fetch all orders for TARGET_DATE and return raw order list."""
    start = f"{DATE_LABEL} 00:00:00"
    end = f"{DATE_LABEL} 23:59:59"

    payload = {
        "dateType": 1,
        "startDatetime": start,
        "endDatetime": end,
        "orderProgressList": ORDER_STATUS_LIST,
    }

    print(f"  Calling Order Search API for {DATE_LABEL} ...")
    try:
        resp = api_post("/2.0/order/searchOrder", payload, auth)
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"Order API HTTP {e.code}: {body}") from e

    error_code = resp.get("errorCode") or resp.get("error_code") or ""
    if error_code:
        raise RuntimeError(f"Order API error: {error_code} - {resp}")

    order_numbers = resp.get("orderNumberList", [])
    if not order_numbers:
        print("  No orders found for the target date.")
        return []

    print(f"  Found {len(order_numbers)} order(s). Fetching details ...")

    # Fetch order details in batches of 100
    orders = []
    batch_size = 100
    for i in range(0, len(order_numbers), batch_size):
        batch = order_numbers[i : i + batch_size]
        detail_resp = api_post(
            "/2.0/order/getOrder",
            {"orderNumberList": batch, "version": 2},
            auth,
        )
        orders.extend(detail_resp.get("orderModelList", []))

    return orders


def aggregate_sales(orders: list[dict]) -> list[dict]:
    """Aggregate orders into per-product sales rows."""
    product_map: dict[str, dict] = {}

    for order in orders:
        for package in order.get("packageModelList", []):
            for item in package.get("itemModelList", []):
                item_id = item.get("itemNumber") or item.get("itemUrl") or "UNKNOWN"
                item_name = item.get("itemName", "")
                unit_price = float(item.get("unitPrice", 0) or 0)
                qty = int(item.get("units", 0) or 0)
                price = float(item.get("price", unit_price * qty) or 0)

                if item_id not in product_map:
                    product_map[item_id] = {
                        "item_number": item_id,
                        "item_name": item_name,
                        "unit_price": unit_price,
                        "quantity_sold": 0,
                        "total_sales_jpy": 0.0,
                        "order_count": 0,
                    }

                product_map[item_id]["quantity_sold"] += qty
                product_map[item_id]["total_sales_jpy"] += price
                product_map[item_id]["order_count"] += 1

    return sorted(
        product_map.values(),
        key=lambda r: r["total_sales_jpy"],
        reverse=True,
    )


def write_sales_csv(rows: list[dict], path: Path) -> None:
    fieldnames = [
        "date",
        "item_number",
        "item_name",
        "unit_price",
        "quantity_sold",
        "total_sales_jpy",
        "order_count",
    ]
    with open(path, "w", newline="", encoding="utf-8-sig") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for row in rows:
            writer.writerow({"date": DATE_LABEL, **row})
    print(f"  Saved: {path}")


# ---------------------------------------------------------------------------
# Traffic extraction (Navigate / Item Stats API)
# ---------------------------------------------------------------------------

def fetch_item_access_counts(auth: str, item_numbers: list[str]) -> dict[str, int]:
    """
    Attempt to retrieve per-item access counts via the Navigate API.

    Rakuten RMS exposes daily access counts through the Shop Navigate endpoint.
    The endpoint returns item-level pageview data when available.
    Returns a dict of {item_number: access_count}.
    """
    access_map: dict[str, int] = {}

    # Try Navigate Top Items endpoint (returns top-accessed items for a date range)
    params = {
        "dateType": "day",
        "startDate": DATE_LABEL,
        "endDate": DATE_LABEL,
        "paginationOffset": 0,
        "paginationLimit": 1000,
    }

    print("  Calling Navigate API for access counts ...")
    try:
        resp = api_get("/1.0/navigate/topNavigateItem", params, auth)
        navigate_items = resp.get("navigateItemModelList", [])
        for entry in navigate_items:
            item_no = entry.get("itemNumber") or entry.get("item_number", "")
            count = int(entry.get("access_count", 0) or entry.get("accessCount", 0))
            if item_no:
                access_map[item_no] = count
        print(f"  Navigate API returned {len(access_map)} item(s).")
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        print(f"  WARNING: Navigate API HTTP {e.code}: {body[:200]}")
    except Exception as exc:
        print(f"  WARNING: Navigate API unavailable: {exc}")

    return access_map


def write_traffic_csv(
    item_numbers: list[str],
    item_names: dict[str, str],
    access_map: dict[str, int],
    path: Path,
) -> None:
    # Merge all known items — union of sales items and navigate items
    all_ids = sorted(set(item_numbers) | set(access_map.keys()))

    fieldnames = ["date", "item_number", "item_name", "access_count", "data_source"]
    with open(path, "w", newline="", encoding="utf-8-sig") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for item_id in all_ids:
            count = access_map.get(item_id)
            source = "Navigate API" if item_id in access_map else "N/A"
            writer.writerow(
                {
                    "date": DATE_LABEL,
                    "item_number": item_id,
                    "item_name": item_names.get(item_id, ""),
                    "access_count": count if count is not None else "",
                    "data_source": source,
                }
            )
    print(f"  Saved: {path}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def write_empty_csv(path: Path, note: str) -> None:
    with open(path, "w", newline="", encoding="utf-8-sig") as f:
        f.write(f"# {note}\n")
        f.write(f"date,note\n")
        f.write(f"{DATE_LABEL},{note}\n")
    print(f"  Saved (empty): {path}")


def main() -> int:
    print(f"=== Rakuten RMS Daily Extract — {DATE_LABEL} ===")
    DATA_DIR.mkdir(parents=True, exist_ok=True)

    sales_path = DATA_DIR / f"rakuten_sales_{DATE_STR}.csv"
    traffic_path = DATA_DIR / f"rakuten_traffic_{DATE_STR}.csv"

    # --- Credentials check ---
    service_secret, license_key = get_credentials()
    if not service_secret or not license_key:
        msg = (
            "RAKUTEN_SERVICE_SECRET and/or RAKUTEN_LICENSE_KEY environment "
            "variables are not set. Skipping API calls."
        )
        print(f"ERROR: {msg}")
        write_empty_csv(sales_path, msg)
        write_empty_csv(traffic_path, msg)
        return 1

    auth = build_auth_header(service_secret, license_key)

    # --- Sales ---
    exit_code = 0
    sales_rows: list[dict] = []
    item_names: dict[str, str] = {}

    print("\n[1/2] Sales (Order API v2)")
    try:
        orders = fetch_orders(auth)
        sales_rows = aggregate_sales(orders)
        item_names = {r["item_number"]: r["item_name"] for r in sales_rows}
        if sales_rows:
            print(f"  Aggregated {len(sales_rows)} product(s).")
            write_sales_csv(sales_rows, sales_path)
        else:
            write_empty_csv(sales_path, "No sales orders found for date")
    except RuntimeError as exc:
        print(f"  ERROR: {exc}")
        write_empty_csv(sales_path, str(exc))
        exit_code = 2

    # --- Traffic ---
    print("\n[2/2] Traffic (Navigate API)")
    item_numbers = [r["item_number"] for r in sales_rows]
    access_map = fetch_item_access_counts(auth, item_numbers)
    write_traffic_csv(item_numbers, item_names, access_map, traffic_path)

    # --- Summary ---
    print("\n=== Summary ===")
    print(f"  Date           : {DATE_LABEL}")
    print(f"  Products sold  : {len(sales_rows)}")
    total_qty = sum(r["quantity_sold"] for r in sales_rows)
    total_rev = sum(r["total_sales_jpy"] for r in sales_rows)
    print(f"  Units sold     : {total_qty}")
    print(f"  Total revenue  : ¥{total_rev:,.0f}")
    print(f"  Products w/ access data: {len(access_map)}")
    print(f"  Sales CSV      : {sales_path}")
    print(f"  Traffic CSV    : {traffic_path}")
    print("=== Done ===")

    return exit_code


if __name__ == "__main__":
    sys.exit(main())
